### PR TITLE
Add an API to choose the underlying JSON encoder.

### PR DESCRIPTION
**Pull request by clokep**
This is an alternative to #24, which allows for better backwards compatibility for a period of time. It lets libraries depending on canonicaljson opt into using a different JSON implementation.

### CODE REVIEW

File: canonicaljson.py [hunk 1]
```diff
@@ -51,19 +40,35 @@ def _default(obj):
...
+
+def set_json_library(json_lib):
+    """
+    Set the encoders for canonicaljson to json_lib.
+
+    Params:
+        json_lib: The module to use for JSON encoding, Must have a
+            `JSONEncoder` property.
+    """
+    globals()["_canonical_encoder"] = json_lib.JSONEncoder(
```
Comment (auscompgeek) [inline canonicaljson.py:58]: Is there a reason why the `global` statement isn't being used?

Comment (clokep) [reply canonicaljson.py:58]: Thanks for your feedback, but I'm still experimenting in this branch. I've reverted this PR back to a draft, which is what I meant to use in the first place.

**Comment by clokep** [comment]
See #26 for the failing builds.

**Comment by clokep** [comment]
This is probably ready to go, minus some CI issues.

**Comment by clokep** [comment]
See matrix-org/synapse#7674 for how this would be used.

File: canonicaljson.py [hunk 2]
```diff
@@ -50,19 +39,37 @@ def _default(obj):
...
+_canonical_encoder = None
+_pretty_encoder = None
+
+
+def set_json_library(json_lib):
+    """
+    Set the underlying JSON library that canonicaljson uses to json_lib.
+
+    Params:
+        json_lib: The module to use for JSON encoding, Must have a
```
Comment (anoadragon453) [inline canonicaljson.py:54]:
```suggestion
        json_lib: The module to use for JSON encoding. Must have a
```

**Comment by clokep** [comment]
@anoadragon453 There's a bunch of conversation about this in https://github.com/matrix-org/synapse/issues/7674#issuecomment-659623838, but the gist is:

* Synapse doesn't put a max version on the canonicaljson dependency, so any change must be backwards compatible to not break a new install of an old Synapse.
* The goal of this is to allow Synapse to *NOT* use simplejson, in a backwards compatible fashion. The approach taken here is that it is backwards compatible unless you opt into new behavior.

> as we seem to use `simplejson` in this library because we think it better performant than stdlib `json`

We used to think this, the latest profiling done shows this to no longer be true.

Let me know if that clears things up or not!

**Comment by anoadragon453** [comment]
Ahh, I see now. We're not only worrying about other users of this library, but also old versions of Synapse. Makes sense :slightly_smiling_face: 



>> as we seem to use simplejson in this library because we think it better performant than stdlib json

> We used to think this, the latest profiling done shows this to no longer be true.

Yes, I was thinking we should just change the default in this library to always be stdlib json based on the new benchmark data, but I'm guessing that would be a backwards incompatible change.

**Comment by clokep** [comment]
> but I'm guessing that would be a backwards incompatible change.

The standard library JSON on Python 3.5 doesn't handle bytes well, so...yeah it isn't backwards compatible. Hopefully we can do that at some point in the future!

**Pull request merged**